migrations/upgrade.ts

```diff
@@ -0,0 +1,185 @@
+                0,
+                messageProxyForMainnetAddress,
+                0,
+                encodedData
+            ));
+
+            safeTransactions.unshift(encodeTransaction(
+                0,
+                marionette.address,
+                0,
+                marionette.interface.encodeFunctionData(
+                    "execute", 
+                    [
+                        proxyAdmin.address,
+                        0,
+                        proxyAdmin.interface.encodeFunctionData("transferOwnership", [multiSend.address])
+                    ]
+                )
+            ));
+
+            safeTransactions.push(encodeTransaction(
+                0,
+                marionette.address,
+                0,
+                marionette.interface.encodeFunctionData(
+                    "execute", 
+                    [
+                        marionette.address,
+                        0,
+                        marionette.interface.encodeFunctionData("revokeRole", [await marionette.PUPPETEER_ROLE(), multiSend.address])
+                    ]
+                )
+            ));
+
+            if (safeMockAddress !== undefined) {
+                safeTransactions.push(encodeTransaction(
+                    0,
+                    proxyAdmin.address,
+                    0,
+                    proxyAdmin.interface.encodeFunctionData("transferOwnership", [safeMockAddress])
+                ));
+            } else {
+                safeTransactions.push(encodeTransaction(
+                    0,
+                    proxyAdmin.address,
+                    0,
+                    proxyAdmin.interface.encodeFunctionData("transferOwnership", [marionette.address])
+                ));
+            }
+ 
+            encodedData = imaMockFactory.interface.encodeFunctionData(
+                "postOutgoingMessage",
+                [
+                    schainHash,
+                    marionette.address,
+                    ethers.utils.defaultAbiCoder.encode(["address", "uint", "bytes"], [
+                        multiSend.address,
+                        0,
+                        multiSendFactory.interface.encodeFunctionData("multiSend", [ concatTransactions(safeTransactions) ])
+                    ])
+                ]
+            );
+
+            // 2.1. transferOwnership from Marionette to MultiSend (for upgrade permissions)
+            // 2.2. upgrade Marionette
+            // 2.3. setVersion
+            // 2.4. revoke PUPPETEER_ROLE from MultiSend
+            // 2.5. transferOwnership from MultiSend to Marionette
+            // or
+            // 2.6. transferOwnership from Marionette to SafeMock
+            _safeTransactions.push(encodeTransaction(
+                0,
+                messageProxyForMainnetAddress,
+                0,
+                encodedData
+            ));
+            safeTransactions.length = 0;
+            Object.assign(safeTransactions, _safeTransactions);
```

Comment:
It's so dirty.
You are provided with a `safeTransactions` object. You create a temporary `_safeTransactions` object. You use `safeTransaction` as auxiliary one and put the result into `_safeTransactions` then copy value from `_safeTransactions` to `safeTransactions`.
In my opinion it's much more clean to create a temporary array for batch 2 transactions and then add the result to `safeTransactions`.